submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,313 @@
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIP. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIP so that remote Pods can communicate
+with it using its globalIP. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIP of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIP.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIPs because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+Egress traffic:
+
+* By default, every cluster will be assigned a configurable number of globalIPs which will be
+  used as egress IPs for any cross-cluster communication. Support for more than a single globalIP
+  at the cluster level is done to avoid ephemeral port exhaustion issues.
+* Provision to allocate a set of globalIPs at the namespace/project level which takes precedence
+  over the globalIPs allocated to the cluster. All Pods in the namespace will use the same globalIPs
+  as EgressIPs.
+* Provision to allocate globalIPs to a Pod (or a set of Pods) from the selected namespaces or
+  from any namespace. Such Pods will use globalIPs that take precedence over any globalIPs allocated
+  at the default namespace level.
+
+Ingress traffic:
+
+* Instead of annotating every ClusterIP Service in the cluster, only exported Services will be
+  annotated with a globalIP.
+* For Headless Services, necessary ingress rules will be programmed to support connectivity to
+  the backend Pods. In a Headless Service use-case, every backend Pod will be annotated with
+  a unique globalIP as an Ingress IP. Clients can discover the associated globalIP using
+  Lighthouse and can connect to the globalIP assigned to the backend pods.
+  For Pods backed by Headless Services, if the user does not explicitly request a globalIP via
+  GlobalnetEgressIP CRD, the same globalIP assigned to the Pods will be used for both Ingress
+  as-well-as Egress. However, if user creates a GlobalnetEgressIP that matches the namespace,
+  or the backend pods of Headless Service, the globalIP assigned via GlobalnetEgressIP takes
+  precedence over individual Pod globalIPs as EgressIP.
```

Comment:
That's for ingress.

For egress, if there  is an specific egress request, it's because there is a necessity for it. So it must take precedence at egress. For example, a remote firewall on a external network expecting static IPs out of the pods, or IP based authentication on a database (a'las mysql), or some sort of debugging happening with the pods over the net.

Otherwise when those pods are upgraded/rotated, the IPs will change, and such authentication / rules / debug will be broken.

The assumption is that if GlobalnetEgressIP is requested over the pods it's a higher priority on the egress.

Of course, this will not take precedence over existing traffic that has been DNATed on ingress over the pods globalnet ingress ips.

Reply:
See the table in here: https://docs.google.com/document/d/1qnZ2LpF_rXGfnYYPNTldQ4WbeEUxwnuQD-xTC6GbZdg/edit#heading=h.t9ki2bstqhx0

Reply:
> This is to support the requirement for ExternalNetwork use-cases where we want a unique IP as EgressIP (even when Pods restart) all the time (if requested by the user).

I agree that requirement of unique IP exists in ExternalNetwork use-cases.
However, is it only with the use cases?

I guess that it would be helpful even in cross-cluster use-cases for users who would like to have a control over IP addresses.